_shared/project/docker-compose.yml

```diff
@@ -1,8 +1,19 @@
-version: "3.9"
 services:
-  web:
-    build: .
+  {% if cookiecutter.get("db") == "yes" %}
+  postgres:
+    image: postgres:{{ cookiecutter.get("__postgres_version", "11.5-alpine") }}
     ports:
-      - "8000:8000"
-  redis:
-    image: "redis:alpine"
+      - "127.0.0.1:{{ cookiecutter['__db_port'] }}:5432"
+    healthcheck:
+      test: ["CMD", "pg_isready", "-U", "postgres"]
+      interval: 1s
+    {% if include_exists("docker-compose/postgres.yml") %}
+      {{- include("docker-compose/postgres.yml", indent=4) -}}
+    {% endif %}
+  {% endif %}
+  {% if include_exists("docker-compose/services.yml") %}
+    {{- include("docker-compose/services.yml", indent=2) -}}
+  {% endif %}
+{% if include_exists("docker-compose/tail.yml") %}
+  {{- include("docker-compose/tail.yml") -}}
+{% endif %}
```

Comment:
> I'd recommend we we change include to check if something exists before do anything, it would remove the need for the ifs.

`include()` actually does [return an empty string if the file doesn't exist](https://github.com/hypothesis/cookiecutters/blob/0c1809aa3c650d3b30f125fdd649b46cd4c596dd/_shared/local_extensions.py#L162), although I should probably remove that feature of `include()` because it's never used anymore.

This can't remove the need for the `{% if %}`'s unfortunately. I think this is similar to our previous discussion about wishing to indent the bodies of Jinja2 blocks: it's just not how Jinja2 works and we're going to have to accept that rather than fight it.

Specifically what's going on here is that a Jinja2 `{% if %}` renders _nothing_ if the expression is false: not a blank line, not a newline character, not whitespace, nothing. Whereas a Jinja2 `{{ expression }}` such as `{{ include("foo") }}` always renders something no matter what the `include()` function returns.

Currently if none of these includes exist then the `{% if %}`'s output nothing at all: not even a blank line, newline character, or any whitespace. The `interval: 1s` line (at the end of the `postgres` service definition) will be the last line in the file. Here's what `docker-compose.yml` looks like when `"db"` is `"yes"` in `cookiecutter.json` but none of the three include files exists:

```yml
services:
  postgres:
    image: postgres:11.5-alpine
    ports:
      - "127.0.0.1:5437:5432"
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "postgres"]
      interval: 1s
```

Here's what that would look like if we removed the `{% if %}`'s and allowed `include()` to return empty strings for include files that don't exist:

```yml
services:
  postgres:
    image: postgres:11.5-alpine
    ports:
      - "127.0.0.1:5437:5432"
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "postgres"]
      interval: 1s
    
  

```

There are three empty lines at the bottom of the file, two of which also have whitespace characters in them.

Here's what it would render if `include()` returned `None` rather than `""` when the file doesn't exist:

```yml
services:
  postgres:
    image: postgres:11.5-alpine
    ports:
      - "127.0.0.1:5437:5432"
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "postgres"]
      interval: 1s
    None
  None
None
```

It actually renders the text `None` into the output file.

You can attempt to use `{{-`'s and `-}}`'s to strip away the unwanted whitespace but that just creates other problems. For example when an include file _does_ exist then `{{-`'s and `-}}`'s will still strip away any whitespace before or after the contents of the include files and run those contents into the lines before and after. There's no way in Jinja2 to say "strip away the unwanted whitespace around this `{{ include() }}`, but only if `include()` didn't return anything".

Trust me, I have tried to do this, and you will drive yourself crazy trying to get it to render correctly for each of the possible combinations of nearby include files existing or not existing or `cookiecutter.json` options being off or on etc. You can't get it right in all cases at once. Nor can you always ignore the problems: a bit of unwanted whitespace _might_ not be meaningful (depending on the output file type) but you'll often get lines and words being run together in ways that will certainly break the output file. The reason for all the trouble is that you're misunderstanding how Jinja2 works and fighting against: once you start using `{% if %}`'s it all just works easily

Reply:
> I've separately recommending that we always include something (e.g. a comment) in the file so you know it exists, so there would be some output and the if would also not be necessary.

We discussed this one in Slack but for the record I don't think this works either. I don't want to litter our output files with `"foo include goes here"` comments. Not all file format allow comments in all locations. And I also don't think it's possible for the cookiecutter to ensure that a default or placeholder include file (containing a comment) always exists for every supported include file: I think there'd be lots of problems with trying to implement that and I don't think it'll ultimately be possible to get it working.

Reply:
> I think we should also be kind of committed to not caring about non-functional whitespace changes.

Mentioned above but it's not only meaningless whitespace that we'd end up with if we started trying to call `{{ include() }}` without `{% if %}`'s. In some cases you end up running lines together. In some output files whitespace is actually meaningful. And in some cases you would get away with it.

I think we should follow the approach of making the output files look nice, neat and readable in terms of formatting, whitespace, blank lines, etc, and eat the cost of the templates being a bit ugly. This is for a few reasons. One is that I just want the files in our dev environments to be nice and neat. Another is that it's not actually possible for the templates themselves to look nice, for a number of reasons:

* Template files are a mix of Jinja2 and YAML, INI, TOML, or whatever the output file format is going to be. There's no way to make this nice: the mixing together of the two completely different syntaxes is just horrible
* As previously discussed you can't indent the bodies of Jinja2 `{% if %}`'s, `{% else %}`s, `{% for %}`'s, macros, etc without breaking the indentation of the output file and indentation is meaningful in most of the cookiecutter's output file formats
* In some cases you won't be able to get rid of `{% if %}`'s from the templates either because this would cause more than just unwanted whitespace (e.g. running lines together) or because whitespace is actually meaningful in the output file

So we could say "we don't care about meaningless whitespace in the output files" and remove some of the `{% if %}`'s from the templates. But this still wouldn't make the templates nice at all: they'd still be a horrible mix of Jinja2 and the output format's syntax, Jinja2 blocks would still not be indented, and in some cases `{% if %}`'s would still be needed. So now we'd have the worst of both worlds: both our templates and our output files would be ugly.

The templates actually _can't_ be pretty but the output files can be.